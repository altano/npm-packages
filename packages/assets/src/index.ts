import { createRequire } from "node:module";
import { readFile } from "node:fs/promises";

const resolve = createRequire(import.meta.url).resolve;

/**
 * @returns the file path for the Inter woff font file at the given weight
 */
export function getInterPath(fontWeight: number): string {
  return resolve(
    `@fontsource/inter/files/inter-latin-${fontWeight}-normal.woff`,
  );
}

/**
 * @returns the node Buffer for the Inter woff font file at the given weight
 */
export async function getInterBuffer(fontWeight: number): Promise<Buffer> {
  return readFile(getInterPath(fontWeight));
}
