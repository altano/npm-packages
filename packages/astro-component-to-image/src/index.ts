import satori from "satori";
import { defineMiddleware } from "astro/middleware";
import sharp, { type Sharp, type FormatEnum as SharpFormatEnum } from "sharp";
import { html as htmlToVNode } from "satori-html";
import { contentType } from "mime-types";
import he from "he";

import type { MiddlewareResponseHandler, APIContext } from "astro";
import type { SatoriOptions as SatoriOptionsOrig } from "satori";

export type SatoriOptions = SatoriOptionsOrig;
export type SharpOptions = NonNullable<Parameters<Sharp["toFormat"]>[1]>;
export type SharpFormats = keyof SharpFormatEnum;

// TODO Needs lots of tests

export type ImageOptions<Format extends SharpFormats> = {
  format: Format;
  options?: SharpOptions;
};

type Filename<Format extends SharpFormats> = `${string}.${Format}`;

export type Options<Format extends SharpFormats> = {
  /**
   * Any output format that the sharp library accepts, as a string.
   * e.g. "avif", "jpg", "png", "webp", "gif", etc.
   *
   * API documentation: https://sharp.pixelplumbing.com/api-output
   */
  format: Format;
  /**
   * Options that the vercel/satori library accepts. At the very least, you must
   * specify dimensions and one font to use.
   *
   * API documentation:
   * https://github.com/vercel/satori/blob/0a258931fe2291bdd461103780ac01e3c700b845/src/satori.ts#L18-L40
   */
  getSatoriOptions(
    context: APIContext,
    response: Response,
    format: Format,
  ): Promise<SatoriOptions>;
  /**
   * Any output options that the sharp library accepts, e.g.:
   *
   *   { quality: 50 }
   *
   * API documentation: https://sharp.pixelplumbing.com/api-output#toformat
   */
  getSharpOptions?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<SharpOptions>;
  /**
   * This function must return true for the given request or the route will not
   * be converted to an image. By default, only components/endpoints that return
   * a text/html Response and have a route ending in "<name>.<format>", e.g.
   * image.png, will be converted to an image. Use this method to override that
   * behavior and provide your own heuristic.
   */
  shouldReplace?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<boolean>;
  /**
   * The default filename for a component such as `image.png.astro` is
   * `image.png`. You may use this function to override that default. Only
   * relevant in SSG: in SSR there is no filename and Astro just responds with
   * the file image contents.
   */
  getFilename?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<Filename<Format>>;
};

function getContentType<Format extends SharpFormats>(format: Format): string {
  const type = contentType(format);
  return type === false ? `image/${format}` : type;
}

async function defaultGetFilename<Format extends SharpFormats>(
  format: Format,
  context: APIContext,
): Promise<string> {
  const requestUrl = context.url.toString();
  new RegExp(String.raw`\.${format}\/?$`).test(requestUrl);

  const basename = context.url.pathname.split("/").filter(Boolean).at(-1);

  if (basename == null) {
    throw new Error(
      `Could not find basename of URL: ${context.url.toString()}`,
    );
  }

  return basename;
}

async function defaultGetSharpOptions(): Promise<SharpOptions> {
  return { quality: 100 };
}

async function defaultShouldReplace<Format extends SharpFormats>(
  context: APIContext,
  response: Response,
  format: Format,
): Promise<boolean> {
  const contentType = response.headers?.get("content-type");

  // In SSG there will be a trailing slash in the request URL. In SSR/dev a
  // user might type it in without the slash. Handle both.
  const requestUrl = context.url.toString();
  const isImageOfType = new RegExp(String.raw`\.${format}\/?$`).test(
    requestUrl,
  );

  return (
    isImageOfType &&
    // Don't rewrite other things like Endpoints. Only interested in components.
    response instanceof Response &&
    // Don't rewrite anything that isn't html.
    contentType === "text/html"
  );
}

export function createHtmlToImageMiddleware<Format extends SharpFormats>({
  format,
  getSharpOptions,
  shouldReplace,
  getFilename,
  getSatoriOptions,
}: Options<Format>): MiddlewareResponseHandler {
  return defineMiddleware(async (context, next) => {
    const response = await next();
    const replace =
      shouldReplace == null
        ? await defaultShouldReplace(context, response, format)
        : await shouldReplace(context, response, format);

    if (!replace) {
      // This request isn't relevant to us. Don't modify the response.

      // TODO Remove when issue is fixed in astro
      // https://github.com/withastro/astro/issues/8045
      if (isEndpointOutput(response)) {
        return new Response(response.body, {
          status: 200,
          headers: response.headers,
        });
      }

      return response;
    }

    const [responseText, satoriOptions, sharpOptions, filename] =
      await Promise.all([
        response.text(),
        getSatoriOptions(context, response, format),
        getSharpOptions == null
          ? defaultGetSharpOptions()
          : getSharpOptions(context, response, format),
        getFilename == null
          ? defaultGetFilename(format, context)
          : getFilename(context, response, format),
      ]);

    // html text => vnode
    const responseTextWithDecodedHtmlEntities = he.decode(responseText);
    const vnode = htmlToVNode(responseTextWithDecodedHtmlEntities);

    // vnode => svg
    const svg = await satori(vnode as React.ReactNode, satoriOptions);

    // svg => image
    const imageBuffer = await sharp(Buffer.from(svg))
      .toFormat(format, sharpOptions)
      .toBuffer();

    return new Response(imageBuffer, {
      headers: {
        "Content-Type": getContentType(format),
        "Content-Disposition": `inline; filename="${filename}"`,
      },
    });
  });
}

function isEndpointOutput(endpointResult: unknown): boolean {
  return (
    endpointResult != null &&
    !(endpointResult instanceof Response) &&
    typeof endpointResult === "object" &&
    "body" in endpointResult &&
    typeof endpointResult.body === "string"
  );
}
