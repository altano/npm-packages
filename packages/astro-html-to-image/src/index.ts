import satori from "satori";
import { defineMiddleware } from "astro/middleware";
import { Resvg } from "@resvg/resvg-wasm";
import { html as htmlToVNode } from "satori-html";
import { contentType } from "mime-types";
import he from "he";

import type { MiddlewareResponseHandler, APIContext } from "astro";
import type { SatoriOptions } from "satori";
import type { ResvgRenderOptions } from "@resvg/resvg-wasm";

export type SvgOptions = SatoriOptions;
export type ImageRenderOptions = ResvgRenderOptions;
export type ImageFormat = "png";

// TODO Needs lots of tests

export type ImageOptions<Format extends ImageFormat> = {
  format: Format;
  options?: ImageRenderOptions;
};

type Filename<Format extends ImageFormat> = `${string}.${Format}`;

export type Runtime = "nodejs";

export async function initialize({
  runtime,
}: {
  runtime: Runtime;
}): Promise<void> {
  switch (runtime) {
    case "nodejs":
      {
        const initialize = await import("./initializeResvg-Nodejs");
        await initialize.default();
      }
      break;
    default:
      runtime satisfies never;
  }
}

export type Options<Format extends ImageFormat> = {
  /**
   * Your JavaScript runtime, e.g. "nodejs"
   */
  runtime: Runtime;
  /**
   * Any output format that @resvg/resvg-wasm library accepts, as a string.
   * Currently only "png"
   */
  format: Format;
  /**
   * Options that the vercel/satori library accepts. At the very least, you must
   * specify dimensions and one font to use.
   *
   * API documentation:
   * https://github.com/vercel/satori/blob/0a258931fe2291bdd461103780ac01e3c700b845/src/satori.ts#L18-L40
   */
  getSvgOptions(
    context: APIContext,
    response: Response,
    format: Format,
  ): Promise<SvgOptions>;
  /**
   * Any output options that the @resvg/resvg-js library accepts, e.g.:
   *
   *   { background: "rgba(255,255,255, 0.8)" }
   *
   * API documentation: https://github.com/yisibl/resvg-js/blob/main/index.d.ts#L3
   */
  getImageOptions?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<ImageRenderOptions>;
  /**
   * This function must return true for the given request or the route will not
   * be converted to an image. By default, only components/endpoints that return
   * a text/html Response and have a route ending in "<name>.<format>", e.g.
   * image.png, will be converted to an image. Use this method to override that
   * behavior and provide your own heuristic.
   */
  shouldReplace?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<boolean>;
  /**
   * The default filename for a component such as `image.png.astro` is
   * `image.png`. You may use this function to override that default. Only
   * relevant in SSG: in SSR there is no filename and Astro just responds with
   * the file image contents.
   */
  getFilename?: (
    context: APIContext,
    response: Response,
    format: Format,
  ) => Promise<Filename<Format>>;
};

function getContentType<Format extends ImageFormat>(format: Format): string {
  const type = contentType(format);
  return type === false ? `image/${format}` : type;
}

export async function defaultGetFilename<Format extends ImageFormat>(
  format: Format,
  context: APIContext,
): Promise<string> {
  const requestUrl = context.url.toString();
  new RegExp(String.raw`\.${format}\/?$`).test(requestUrl);

  const basename = context.url.pathname.split("/").filter(Boolean).at(-1);

  if (basename == null) {
    throw new Error(
      `Could not find basename of URL: ${context.url.toString()}`,
    );
  }

  return basename;
}

export async function defaultGetImageOptions(): Promise<ImageRenderOptions> {
  return {
    textRendering: 1, // optimizeLegibility
    imageRendering: 0, // optimizeQuality
  };
}

export async function defaultShouldReplace<Format extends ImageFormat>(
  context: APIContext,
  response: Response,
  format: Format,
): Promise<boolean> {
  const contentType = response.headers?.get("content-type");

  // In SSG there will be a trailing slash in the request URL. In SSR/dev a
  // user might type it in without the slash. Handle both.
  const requestUrl = context.url.toString();
  const isImageOfType = new RegExp(String.raw`\.${format}\/?$`).test(
    requestUrl,
  );

  return (
    isImageOfType &&
    // Don't rewrite other things like Endpoints. Only interested in components.
    response instanceof Response &&
    // Don't rewrite anything that isn't html.
    contentType === "text/html"
  );
}

let isWasmInitialized = false;

export function createHtmlToImageMiddleware<Format extends ImageFormat>({
  runtime,
  format,
  getImageOptions,
  shouldReplace,
  getFilename,
  getSvgOptions,
}: Options<Format>): MiddlewareResponseHandler {
  return defineMiddleware(async (context, next) => {
    if (!isWasmInitialized) {
      isWasmInitialized = true;
      try {
        await initialize({ runtime });
      } catch (err) {
        // Restarting the dev server causes us to unnecessarily initialize the
        // wasm module twice. Let's just ignore the error (since we can't
        // prevent it).
        const shouldIgnore =
          err instanceof Error &&
          err.message?.startsWith("Already initialized");
        if (!shouldIgnore) {
          // re-throw all other errors
          throw err;
        }
      }
    }

    const response = await next();
    const replace =
      shouldReplace == null
        ? await defaultShouldReplace(context, response, format)
        : await shouldReplace(context, response, format);

    if (!replace) {
      // This request isn't relevant to us. Don't modify the response.

      // TODO Remove when issue is fixed in astro
      // https://github.com/withastro/astro/issues/8045
      if (isEndpointOutput(response)) {
        return new Response(response.body, {
          status: 200,
          headers: response.headers,
        });
      }

      return response;
    }

    const [responseText, svgOptions, imageOptions, filename] =
      await Promise.all([
        response.text(),
        getSvgOptions(context, response, format),
        getImageOptions == null
          ? defaultGetImageOptions()
          : getImageOptions(context, response, format),
        getFilename == null
          ? defaultGetFilename(format, context)
          : getFilename(context, response, format),
      ]);

    // html text => vnode
    const responseTextWithDecodedHtmlEntities = he.decode(responseText);
    const vnode = htmlToVNode(responseTextWithDecodedHtmlEntities);

    // vnode => svg
    const svg = await satori(vnode as React.ReactNode, svgOptions);

    // svg => image
    const imageResvgArr = new Resvg(svg, imageOptions).render().asPng();
    const imageBuffer = Buffer.from(imageResvgArr);

    return new Response(imageBuffer, {
      headers: {
        "Content-Type": getContentType(format),
        "Content-Disposition": `inline; filename="${filename}"`,
      },
    });
  });
}

function isEndpointOutput(endpointResult: unknown): boolean {
  return (
    endpointResult != null &&
    !(endpointResult instanceof Response) &&
    typeof endpointResult === "object" &&
    "body" in endpointResult &&
    typeof endpointResult.body === "string"
  );
}
