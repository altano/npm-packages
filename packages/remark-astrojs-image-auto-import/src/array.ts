export function extract<T, TExtracted extends T>(
  arr: T[],
  guard: (item: T) => item is TExtracted,
): [extracted: TExtracted | null, rest: T[]] {
  const extractedTest = arr.find(guard);
  if (extractedTest == null) {
    return [null, arr];
  }
  // A little wasteful but *shrug*
  const extractedIndex = arr.findIndex((i) => i === extractedTest);
  const rest = [
    ...arr.slice(0, extractedIndex),
    ...arr.slice(extractedIndex + 1),
  ];
  return [extractedTest, rest];
}
