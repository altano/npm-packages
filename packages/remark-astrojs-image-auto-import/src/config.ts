// import type { Element } from "hast";
export interface RemarkAstroJSImageAutoImportOptions {
  /**
   * When an attribute points at a file that isn't present on the local
   * filesystem an exception will be thrown by default. This option disables
   * that behavior and causes the file to be skipped.
   *
   * @default false
   */
  ignoreFileNotFound?: boolean;
  /**
   * Skip when an attribute points to an HTTP URL that isn't local
   *
   * @default true
   */
  ignoreNonFileUrl?: boolean;
}

export type RemarkAstroJSImageAutoImportConfig =
  Required<RemarkAstroJSImageAutoImportOptions>;

const configDefaults: RemarkAstroJSImageAutoImportConfig = {
  ignoreFileNotFound: false,
  ignoreNonFileUrl: true,
};

export function getConfig(
  options: RemarkAstroJSImageAutoImportOptions,
): RemarkAstroJSImageAutoImportConfig {
  return Object.assign(configDefaults, options);
}
