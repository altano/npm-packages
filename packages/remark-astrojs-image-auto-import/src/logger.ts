import { Logger } from "@altano/remark-plugin-helpers";

const logger = new Logger("@altano/remark-astrojs-image-auto-import");

export default logger;
