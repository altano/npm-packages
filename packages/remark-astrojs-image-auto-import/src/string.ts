export function unquoteIfNecessary(str: string): string {
  if (str.startsWith(`"`) && str.endsWith(`"`)) {
    return str.slice(1, -1);
  }
  return str;
}
