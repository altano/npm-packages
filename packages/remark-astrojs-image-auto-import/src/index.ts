import { is } from "unist-util-is";
import { visitAndReplace } from "./mdx/visit";
import { stringSrcToImportSrc } from "./mdx/srcStringToImport";
import { getConfig } from "./config";

import type { VFile } from "vfile";
import type { MdxJsxFlowElement } from "mdast-util-mdx-jsx";
import type { Plugin } from "unified";
import type { Data, Node } from "unist";
import type { Root } from "mdast";
import type {
  RemarkAstroJSImageAutoImportConfig,
  RemarkAstroJSImageAutoImportOptions,
} from "./config";

interface AstroJSImageElement extends MdxJsxFlowElement {
  name: "Image" | "Picture";
}

function isAstroJSImageElement(node: Node<Data>): node is AstroJSImageElement {
  return (
    is(node, { type: "mdxJsxFlowElement", name: "Image" }) ||
    is(node, { type: "mdxJsxFlowElement", name: "Picture" })
  );
}

function isIgnoredNode(image: AstroJSImageElement): boolean {
  return image.attributes.some((attr) => {
    return (
      attr.type === "mdxJsxAttribute" && attr.name === "data-import-ignore"
    );
  });
}

const transformer = async (
  config: RemarkAstroJSImageAutoImportConfig,
  tree: Root,
  vfile: VFile,
): Promise<void> => {
  await visitAndReplace(tree, async (node) => {
    if (!isAstroJSImageElement(node) || isIgnoredNode(node)) {
      return;
    }
    return stringSrcToImportSrc(node, vfile, config);
  });
};

/**
 * A Remark plugin for converting the src attribute of @astrojs/image components
 * to inline imports
 */
const remarkAstroJSImageAutoImportPlugin: Plugin<
  [RemarkAstroJSImageAutoImportOptions?],
  Root
> = (options) => {
  const config = getConfig(options ?? {});
  return transformer.bind(null, config);
};

export default remarkAstroJSImageAutoImportPlugin;
