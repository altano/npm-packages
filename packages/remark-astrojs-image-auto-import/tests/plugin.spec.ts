import plugin from "../src/index";
import { describe } from "vitest";
import { testByFixtures } from "@altano/remark-plugin-test-util/src/testByFixture";

describe("remark-astrojs-image-auto-import", async () => {
  await testByFixtures(plugin);
});
