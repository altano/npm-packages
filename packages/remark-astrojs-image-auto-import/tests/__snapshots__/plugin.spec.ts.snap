// Vitest Snapshot v1

exports[`remark-astrojs-image-auto-import > [test #0] alt 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
/*IGNORED (only Image components processed)*/
/*IGNORED (only Image components processed)*/
/*IGNORED (src is not a string)*/
/*IGNORED (src is not a string)*/
/*TRANSFORMED*/
/*TRANSFORMED*/
/*TRANSFORMED*/
/*TRANSFORMED*/
/*TRANSFORMED*/
import { Image } from "@astrojs/image/components";
import image from "./image.png";
function _createMdxContent(props) {
  const _components = Object.assign(
    {
      p: "p",
      img: "img",
    },
    props.components,
  );
  return (
    <>
      {}
      {"\\n"}
      <_components.p>
        <_components.img src="image.png" alt="Alt text" />
      </_components.p>
      {"\\n"}
      {}
      {"\\n"}
      <img src={image} />
      {"\\n"}
      {}
      {"\\n"}
      <Image src={image} />
      {"\\n"}
      {}
      {"\\n"}
      <Image src={import("./image.png")} />
      {"\\n"}
      {}
      {"\\n"}
      <Image
        src={import(
          "<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/alt/image.png"
        )}
      />
      {"\\n"}
      {}
      {"\\n"}
      <Image
        src={import(
          "<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/alt/image.png"
        )}
      />
      {"\\n"}
      {}
      {"\\n"}
      <Image
        width={300}
        src={import(
          "<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/alt/image.png"
        )}
      />
      {"\\n"}
      {}
      {"\\n"}
      <Image
        width={300}
        height={600}
        src={import(
          "<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/alt/image.png"
        )}
      />
      {"\\n"}
      {}
      {"\\n"}
      <Image
        aspectRatio="16:9"
        format="avif"
        src={import(
          "<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/alt/image.png"
        )}
      />
    </>
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-auto-import > [test #1] basic 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
import { Image } from "@astrojs/image/components";
function _createMdxContent(props) {
  return (
    <Image
      src={import(
        "<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/basic/image.png"
      )}
    />
  );
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-auto-import > [test #2] data-import-ignore 1`] = `
/*@jsxRuntime automatic @jsxImportSource react*/
import { Image } from "@astrojs/image/components";
function _createMdxContent(props) {
  return <Image data-import-ignore src="./image.png" />;
}
function MDXContent(props = {}) {
  const { wrapper: MDXLayout } = props.components || {};
  return MDXLayout ? (
    <MDXLayout {...props}>
      <_createMdxContent {...props} />
    </MDXLayout>
  ) : (
    _createMdxContent(props)
  );
}
export default MDXContent;

`;

exports[`remark-astrojs-image-auto-import > [test #3] error-imagenotfound 1`] = `"<repo-root>/packages/remark-astrojs-image-auto-import/tests/__fixtures__/error-imagenotfound/non-existent.png could not be found on disk and the \\"ignoreFileNotFound\\" option is false"`;
